src/VisualStudio/Core/Def/Implementation/ProjectSystem/MiscellaneousFilesWorkspace.cs

```diff
@@ -237,7 +237,15 @@ private void Registration_WorkspaceChanged(object sender, EventArgs e)
         {
             var workspaceRegistration = (WorkspaceRegistration)sender;
             uint docCookie;
-            Contract.ThrowIfFalse(_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie));
+
+            // external workspace that listens to events from IVSRunningDocumentTable4 (just like MiscellaneousFilesWorkspace does) and registers the text buffer for opened document prior to Miscellaneous workspace getting notified of the docCookie - the prior contract assumes that MiscellaneousFilesWorkspace is always the first one to get notified
+            if (!_docCookieToWorkspaceRegistration.TryGetKey(workspaceRegistration, out docCookie))
+            {
+                // We haven't even started tracking the document corresponding to this registration - likely because some external workspace registered the document's buffer prior to MiscellaneousWorkspace getting notified of it.
```

Comment:
Note, I have hit this multiple times while debugging my demo prototype (once in 5-6 attempts) and confirmed that everytime that contract failed the below 2 threads were active:

**Thread 1:**
>  	System.Collections.Immutable.dll!System.Collections.Immutable.SortedInt32KeyNode<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket>.MakeBalanced(System.Collections.Immutable.SortedInt32KeyNode<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket> tree)	Unknown
 	System.Collections.Immutable.dll!System.Collections.Immutable.SortedInt32KeyNode<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket>.SetOrAdd(int key, System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket value, System.Collections.Generic.IEqualityComparer<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket> valueComparer, bool overwriteExistingValue, out bool replacedExistingValue, out bool mutated)	Unknown
 	System.Collections.Immutable.dll!System.Collections.Immutable.SortedInt32KeyNode<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket>.SetItem(int key, System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket value, System.Collections.Generic.IEqualityComparer<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket> valueComparer, out bool replacedExistingValue, out bool mutated)	Unknown
 	System.Collections.Immutable.dll!System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.UpdateRoot(System.Collections.Immutable.SortedInt32KeyNode<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket> root, int hashCode, System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket newBucket, System.Collections.Generic.IEqualityComparer<System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.HashBucket> hashBucketComparer)	Unknown
 	System.Collections.Immutable.dll!System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.Add(uint key, Microsoft.CodeAnalysis.WorkspaceRegistration value, System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.KeyCollisionBehavior behavior, System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.MutationInput origin)	Unknown
 	System.Collections.Immutable.dll!System.Collections.Immutable.ImmutableDictionary<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.Add(uint key, Microsoft.CodeAnalysis.WorkspaceRegistration value)	Unknown
 	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.BidirectionalMap<uint, Microsoft.CodeAnalysis.WorkspaceRegistration>.Add(uint key, Microsoft.CodeAnalysis.WorkspaceRegistration value) Line 77	C#
>	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.TrackOpenedDocument(uint docCookie, string moniker) Line 227	C#
 	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.OnAfterAttributeChangeEx(uint docCookie, uint grfAttribs, Microsoft.VisualStudio.Shell.Interop.IVsHierarchy pHierOld, uint itemidOld, string pszMkDocumentOld, Microsoft.VisualStudio.Shell.Interop.IVsHierarchy pHierNew, uint itemidNew, string pszMkDocumentNew) Line 161	C#


**Thread 2:**
>  	Microsoft.VisualStudio.LanguageServices.dll!Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem.MiscellaneousFilesWorkspace.Registration_WorkspaceChanged(object sender, System.EventArgs e) Line 247	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.WorkspaceRegistration.RaiseEvents() Line 31	C#
>	Microsoft.CodeAnalysis.Workspaces.dll!Microsoft.CodeAnalysis.Workspace.RegisterText.AnonymousMethod__0() Line 45	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.TaskExtensions.ContinueWithAfterDelay.AnonymousMethod__1(System.Threading.Tasks.Task _) Line 241	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.TaskExtensions.SafeContinueWith.AnonymousMethod__0(System.Threading.Tasks.Task antecedent) Line 80	C#
 	Microsoft.CodeAnalysis.Workspaces.dll!Roslyn.Utilities.TaskExtensions.SafeContinueWith.AnonymousMethod__0(System.Threading.Tasks.Task t) Line 148	C#
 	mscorlib.dll!System.Threading.Tasks.ContinuationResultTaskFromTask<bool>.InnerInvoke()	Unknown